### ISSUE TITLE
Source catalog step bkg_boxsize keyword type is inconsistent

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#6299]
_Issue [JP-2260](https://jira.stsci.edu/browse/JP-2260) was created on JIRA by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert):_

There is an inconsistency in the bkg_boxsize input parameter of the source_catalog step. The [documentation](https://jwst-pipeline.readthedocs.io/en/latest/jwst/source_catalog/arguments.html) lists this as a floating point number. However, in my test case, when I provide a float, the step crashes. Instead, if I provide bkg_boxsize as a string, the step completes successfully. So it seems like either readthedocs or the code need to be updated.

 
```java
for bsize in np.arange(20, 160, 20):
    ...:     step = SourceCatalogStep()
    ...:     step.bkg_boxsize = 30.
    ...:     step.kernel_fwhm = 2.
    ...:     step.snr_threshold = 50
    ...:     step.npixels = 30
    ...:     step.deblend = True
    ...:     step.save_results = True
    ...:     step.output_file = 'test_bsize_{}'.format(bsize)
    ...:     step.run(file)
    ...: 
2021-08-18 21:00:36,531 - stpipe.SourceCatalogStep - INFO - SourceCatalogStep instance created.
2021-08-18 21:00:36,660 - stpipe.SourceCatalogStep - INFO - Step SourceCatalogStep running with args ('Pipeline_Level2/jw01068004001_01101_00001_nrcb5_cal.fits',).
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
<ipython-input-55-6c68de8eb5dd> in <module>
      8     step.save_results = True
      9     step.output_file = 'test_bsize_{}'.format(bsize)
---> 10     step.run(file)
     11 


~/miniconda3/envs/jwst-py3.8/lib/python3.8/site-packages/stpipe/step.py in run(self, *args)
    370 
    371         self.log.info(
--> 372             f'Step {self.name} parameters are: {self.get_pars()}'
    373         )
    374 


~/miniconda3/envs/jwst-py3.8/lib/python3.8/site-packages/stpipe/step.py in get_pars(self, full_spec)
   1205                 value = getattr(self, key)
   1206                 instance_pars[key] = value
-> 1207         pars = config_parser.config_from_dict(instance_pars, spec, allow_missing=True)
   1208 
   1209         # Convert the config to a pure dict.


~/miniconda3/envs/jwst-py3.8/lib/python3.8/site-packages/stpipe/config_parser.py in config_from_dict(d, spec, root_dir, allow_missing)
    272 
    273     if spec:
--> 274         validate(config, spec, root_dir=root_dir, allow_missing=allow_missing)
    275     else:
    276         config.walk(string_to_python_type)


~/miniconda3/envs/jwst-py3.8/lib/python3.8/site-packages/stpipe/config_parser.py in validate(config, spec, section, validator, root_dir, allow_missing)
    361 
    362         if len(messages):
--> 363             raise ValidationError('\n'.join(messages))
    364     finally:
    365         config.main.configspec = orig_configspec


ValidationError: Config parameter 'bkg_boxsize': the value "30.0" is of the wrong type.


In [56]: for bsize in np.arange(20, 160, 20):
    ...:     step = SourceCatalogStep()
    ...:     step.bkg_boxsize = str(bsize)
    ...:     step.kernel_fwhm = 2.
    ...:     step.snr_threshold = 50
    ...:     step.npixels = 30
    ...:     step.deblend = True
    ...:     step.save_results = True
    ...:     step.output_file = 'test_bsize_{}'.format(bsize)
    ...:     step.run(file)
    ...: 
2021-08-18 21:10:00,645 - stpipe.SourceCatalogStep - INFO - SourceCatalogStep instance created.
2021-08-18 21:10:00,764 - stpipe.SourceCatalogStep - INFO - Step SourceCatalogStep running with args ('Pipeline_Level2/jw01068004001_01101_00001_nrcb5_cal.fits',).
2021-08-18 21:10:00,931 - stpipe.SourceCatalogStep - INFO - Step SourceCatalogStep parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': '/System/Volumes/Data/ifs/jwst/wit/nircam/simulationWG/Imaging/CAR-19/2021_June/test_bsize_20', 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': True, 'skip': False, 'suffix': 'cat', 'search_output_file': True, 'input_dir': '', 'bkg_boxsize': 20, 'kernel_fwhm': 2.0, 'snr_threshold': 50, 'npixels': 30, 'deblend': True, 'aperture_ee1': 30, 'aperture_ee2': 50, 'aperture_ee3': 70, 'ci1_star_threshold': 2.0, 'ci2_star_threshold': 1.8}
2021-08-18 21:10:02,972 - CRDS - INFO -  Fetching  /Users/hilbert/crds_cache/references/jwst/nircam/jwst_nircam_apcorr_0004.fits   20.2 K bytes  (1 / 1 files) (0 / 20.2 K bytes)
2021-08-18 21:10:03,163 - stpipe.SourceCatalogStep - INFO - Using APCORR reference file /Users/hilbert/crds_cache/references/jwst/nircam/jwst_nircam_apcorr_0004.fits
2021-08-18 21:10:03,641 - stpipe.SourceCatalogStep - INFO - Using ABVEGAOFFSET reference file /Users/hilbert/crds_cache/references/jwst/nircam/jwst_nircam_abvegaoffset_0001.asdf
2021-08-18 21:10:03,642 - stpipe.SourceCatalogStep - INFO - Instrument: NIRCAM
2021-08-18 21:10:03,642 - stpipe.SourceCatalogStep - INFO - Detector: NRCBLONG
2021-08-18 21:10:03,642 - stpipe.SourceCatalogStep - INFO - Filter: F356W
2021-08-18 21:10:03,642 - stpipe.SourceCatalogStep - INFO - Pupil: CLEAR
2021-08-18 21:10:03,642 - stpipe.SourceCatalogStep - INFO - Subarray: FULL
2021-08-18 21:10:03,693 - stpipe.SourceCatalogStep - WARNING - model.meta.resample.pixel_scale_ratio was not found. Assuming the native detector pixel scale (i.e., pixel_scale_ratio = 1)
2021-08-18 21:10:03,726 - stpipe.SourceCatalogStep - INFO - AB to Vega magnitude offset 2.82390
2021-08-18 21:10:03,748 - stpipe.SourceCatalogStep - WARNING - There are no valid pixels. Source catalog will not be created.
2021-08-18 21:10:03,748 - stpipe.SourceCatalogStep - INFO - Step SourceCatalogStep done
```

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2260?focusedCommentId=545372#comment-545372):_

The step class actually defines `bkg_boxsize` as integer:
<https://github.com/spacetelescope/jwst/blob/master/jwst/source_catalog/source_catalog_step.py#L31>

where did you see it described as float? In readthedocs?

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2260?focusedCommentId=545381#comment-545381):_

Yes. See the documentation link in the description. That goes to readthedocs, where it specifies a float.

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2260?focusedCommentId=545382#comment-545382):_

Very strange that it works with a string.

I just tried again with an integer and it appeared like it was going to work. It failed, but I believe for an unrelated reason.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2260?focusedCommentId=545385#comment-545385):_

Tracing bkg_boxsize through the code shows that at some point the user input value goes through `astype(int)`, which I assume should work for both float and string inputs, but apparently not. Or something else is going wrong before that point, perhaps already when the input values are parsed at the beginning of the step.

**Comment by stscijgbot-jp** [comment]
_Comment by [Larry Bradley](https://jira.stsci.edu/secure/ViewProfile.jspa?name=lbradley) on [JIRA](https://jira.stsci.edu/browse/JP-2260?focusedCommentId=545399#comment-545399):_

bkg_boxsize should be an integer ([https://github.com/spacetelescope/jwst/blob/master/jwst/source_catalog/source_catalog_step.py#L31)] – I'll update the docs.  However, it should work in principal with a float because of the int(bkg_boxsize).  I'll take a look.

**Comment by stscijgbot-jp** [comment]
_Comment by [Larry Bradley](https://jira.stsci.edu/secure/ViewProfile.jspa?name=lbradley) on [JIRA](https://jira.stsci.edu/browse/JP-2260?focusedCommentId=545403#comment-545403):_

Indeed, this line <https://github.com/spacetelescope/jwst/blob/master/jwst/source_catalog/source_catalog.py#L298>

will cast `bkg_boxsize` into an integer.  So floats and "integer strings" will actually work.  I'm not sure what to fix here, but to be safe I could raise a TypeError if `bkg_boxsize` is not input as an integer.

**Comment by larrybradley** [comment]
Actually, `stpipe` appears to be doing some data type validation (that's why the float inputs fail) based on the `bkg_boxsize` `spec` type (which is integer).  Not sure why it doesn't catch strings, but there may be a bug in `stpipe`.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2260?focusedCommentId=545490#comment-545490):_

Fixed by #6300
